nexus/src/app/sagas/instance_ip_attach.rs

```diff
@@ -0,0 +1,527 @@
+                    external_ip: Some(external_ip),
+                    do_saga,
+                })
+        }
+    }
+}
+
+async fn siia_begin_attach_ip_undo(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let log = sagactx.user_data().log();
+    warn!(log, "siia_begin_attach_ip_undo: Reverting detached->attaching");
+    let params = sagactx.saga_params::<Params>()?;
+    if !instance_ip_move_state(
+        &sagactx,
+        &params.serialized_authn,
+        IpAttachState::Attaching,
+        IpAttachState::Detached,
+    )
+    .await?
+    {
+        error!(log, "siia_begin_attach_ip_undo: external IP was deleted")
+    }
+
+    Ok(())
+}
+
+async fn siia_get_instance_state(
+    sagactx: NexusActionContext,
+) -> Result<Option<Uuid>, ActionError> {
+    let params = sagactx.saga_params::<Params>()?;
+    instance_ip_get_instance_state(
+        &sagactx,
+        &params.serialized_authn,
+        &params.authz_instance,
+        "attach",
+    )
+    .await
+}
+
+async fn siia_nat(sagactx: NexusActionContext) -> Result<(), ActionError> {
+    let params = sagactx.saga_params::<Params>()?;
+    instance_ip_add_nat(
+        &sagactx,
+        &params.serialized_authn,
+        &params.authz_instance,
+    )
+    .await
+}
+
+async fn siia_nat_undo(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let log = sagactx.user_data().log();
+    let params = sagactx.saga_params::<Params>()?;
+    if let Err(e) = instance_ip_remove_nat(
+        &sagactx,
+        &params.serialized_authn,
+        &params.authz_instance,
+    )
+    .await
+    {
+        error!(log, "siia_nat_undo: failed to notify DPD: {e}");
+    }
+
+    Ok(())
+}
+
+async fn siia_update_opte(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let params = sagactx.saga_params::<Params>()?;
+    instance_ip_add_opte(&sagactx, &params.authz_instance).await
+}
+
+async fn siia_update_opte_undo(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let log = sagactx.user_data().log();
+    let params = sagactx.saga_params::<Params>()?;
+    if let Err(e) =
+        instance_ip_remove_opte(&sagactx, &params.authz_instance).await
+    {
+        error!(log, "siia_update_opte_undo: failed to notify sled-agent: {e}");
+    }
+    Ok(())
+}
+
+async fn siia_complete_attach(
+    sagactx: NexusActionContext,
+) -> Result<views::ExternalIp, ActionError> {
+    let log = sagactx.user_data().log();
+    let params = sagactx.saga_params::<Params>()?;
+    let target_ip = sagactx.lookup::<ModifyStateForExternalIp>("target_ip")?;
+
+    if !instance_ip_move_state(
+        &sagactx,
+        &params.serialized_authn,
+        IpAttachState::Attaching,
+        IpAttachState::Attached,
+    )
+    .await?
+    {
+        warn!(
+            log,
+            "siia_complete_attach: external IP was deleted or call was idempotent"
+        )
```

Comment:
I had intentionally designed for Case 2, but had partly overlooked Case 1 (assuming that both are stop->delete) -- thanks a lot for catching that. I agree on that unwind pathway; although I'd (maybe incorrectly) assumed that the instance would be guaranteed to be removed from InstanceManager before moving from `Stopping->Stopped`, since any OPTE endpoint on sled-agent will then give us `Error::NoSuchInstance`.

I'd also be happier to unwind here rather than to push the limits of what we can get away with. Examining it a bit more closely however, there is another issue with the Case 1 race. The NAT entry will not be removed on unwind -- it's no longer attached to the instance, so `instance_delete_dpd_config` will not see it. We need a tweak there to handle this and not wipe out a potential followup attach on another instance.

Reply:
Okay, I've resolved this by unwinding if we observe a delete at the end of the attach saga. I've put in some extra work to clean up straggler NAT entries more precisely (needed in the delete-unwind case) and to make `instance_ensure_dpd_config` more aggressive in pushing routes where an old conflict may exist. I think doing both of these means more useful buffers between us and catastrophe.